### PR TITLE
Improve logic for parsing sota toml configuration

### PR DESCRIPTION
**Pull request by doanac**

### CODE REVIEW

File: internal/app.go
Comment (StealthyCoder) [inline internal/app.go:186]: Do we need to verify anywhere if this path/location actually exists and is writable? Since we before this actually checked if `SOTA_DIR` was a directory.

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:13]: One of the API ctor's allows customer to set custom list of config dirs or files.
So, this won't work if the custom client's list of conf dirs differs from the `DEF_CONFIG_ORDER`. 

It's not critical at the moment, however can become critical as more customers build custom sota client.

Comment (doanac) [reply internal/app_config.go:13]: Sounds like we'll need a configuration knob for this configuration knob. Or - you can probably update the systemd service unit to pass the config dirs you want to the daemon's CLI

File: internal/app.go
Comment (doanac) [reply internal/app.go:186]: Good question. I don't think we were checking before, but still - what's the best way to fail? Here's some thoughts:

* If this path doesn't exist - aklite won't be able function.
* This is something that would happen *before* going to production
* Fioconfig is still somewhat functional even without this directory

Given those points - Its probably worth just trying to run and maybe someone can trigger an OTA via fioconfig that could fix the device

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:46]: Not sure about it, but maybe we should skip links?

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:63]: Is this how the merge occurs, through the assignment cfg.tree (current tree) = toml.LoadFile(cfg.path) (specific file tree)?

Comment (mike-sul) [reply internal/app_config.go:63]: Ah, no merge, sorry.

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:92]: It looks like `Get` and `GetDefault` can be merged into a single `Get(key string, default ...string). Not critical though.

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:100]: This write won't work if the most significant file is located in the `/usr` (readonly fs).

Comment (mike-sul) [reply internal/app_config.go:100]: Maybe appconfig should read/get an overall/merged configuration while allow writing only to a specific file(s). If it happens that the value to be set is defined in some more significant and read only config file then just return an error.

File: internal/app.go
Comment (StealthyCoder) [reply internal/app.go:186]: Maybe somewhere just notify it? Either a WARN or something else that makes a developer/person go "Oops", then fix it in the flow of going to production. :+1:

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:92]: I'm not a big fan of golang functions that do that

File: internal/app_config.go
Comment (mike-sul) [reply internal/app_config.go:13]: The dir/file list can be set at a factory level in some file under `/usr`; a default yocto recipe defines it the default list; a customer can override it. Then, both aklite/custom-client and fioconfig reads it from that file.
?

Comment (doanac) [reply internal/app_config.go:13]: maybe - or maybe we just state that this override is only intended for unit tests. production should the default path

Comment (mike-sul) [reply internal/app_config.go:13]: I guess it makes sense, no value in making things more complex. Anyway those defaults dirs are so generic that they are suitable for all use-cases.

If so, we probably need to add a comment about it to this ctor https://github.com/foundriesio/aktualizr-lite/blob/b2140a50e48537680f5832141141d6db579c6e00/include/aktualizr-lite/api.h#L194.

Or maybe remove it at all and just use `cmdline_args["config"]` and this ctor https://github.com/foundriesio/aktualizr-lite/blob/b2140a50e48537680f5832141141d6db579c6e00/include/aktualizr-lite/api.h#L203 for the unit tests and the custom sota example.

Comment (doanac) [reply internal/app_config.go:13]: I think a comment is probably sufficient. I think Patti may have introduced this line for their tests - so just don't do anything that might accidentally break them

Comment (mike-sul) [reply internal/app_config.go:13]: It's funny how my comment to fioconfig code returned back to me as an action item over aklite :).

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:46]: libaktualizr doesn't seem to:
https://github.com/uptane/aktualizr/blob/master/src/libaktualizr/config/base_config.cc#L27

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:100]: There is only one place we use this API: updating pkey/client-cert when doing a rotation. I thought about trying to do a specific file, but the point of this change was kind of because of the fact we don't know which file they might use. This can only work if the file is /var/sota/sota.toml or lives under /etc.

Maybe there's a better approach than creating a generic "write" method?

File: internal/app_config.go
Comment (vkhoroz) [reply internal/app_config.go:46]: I'm on the fence here.

My original comment was going to read like this:
> I'd check for `IsRegular` here.
> Every month there are a plenty of new CVEs about privilege escalation vulnerabilities found in software following symlinks.
> So, unless we have a _strong reason_ to support symlinks here, I'd not support them.

But then I thought: hey, this process is needs access to system files, so what privilege escalation is possible.

So, I'd leave it up to you here.

But, bear in mind, that if you want to support symlinks here - then you need to also support symlinks at line 41 for consistency.

File: internal/app_config.go
Comment (vkhoroz) [inline internal/app_config.go:59]: So, should `/var/sota/conf.d/a-50-sota.toml` go before `/var/sota/sota.toml`, and `/var/sota/conf-d/z-50-sota.toml` after `/var/sota/sota.toml`?

This is what would happen if I'm reading this correctly.

I was always under impression that the `/var/sota/sota.toml` should be the first file to parse.

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:46]: Its not so much I want to support symlinks, but if we are going to change this - then we also need to change libaktuzlizr

File: internal/app_config.go
Comment (vkhoroz) [reply internal/app_config.go:100]: Foremost, @doanac I'm fine with your justification.

Unrelated to this specific change, but related in general.

I guess some of our users would want to put their `sota.toml` of a read-only partition.
For them, we still need to provide an ability to rotate the client certificate.

One such way is to store the sota config client cert paths in a separate file.
So, maybe, we need to consider making that an option in e.g. `lmp-device-register`?

File: internal/app_config.go
Comment (vkhoroz) [inline internal/app_config.go:91]: Nitpick: can be part of the first commit, right?

**Review by vkhoroz — APPROVED** [review]
Looks clean.

I'm a bit bothered by that `Write` thing... but it seems the most naive solution which is good enough.

File: internal/app_config.go
Comment (vkhoroz) [reply internal/app_config.go:46]: If you don't want to change this - then you need to change line 41 above; as the symlink is not supported out there.

File: internal/app_config.go
Comment (mike-sul) [reply internal/app_config.go:100]: My understanding is that even the current `Write` implementation will return an error if a client tries to set a value stored in the read-only partition anyway. So, it kind of already works in the way I described above.

File: internal/app_config.go
Comment (mike-sul) [reply internal/app_config.go:59]: >This is what would happen if I'm reading this correctly.

Yes, this is the way libaktualizr parses the config files.

**Review by mike-sul — APPROVED** [review]
LGTM.

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:91]: > I guess some of our users would want to put their sota.toml of a read-only partition.

The sota.toml itself created by lmp-dr pretty much has to be writable.

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:46]: https://github.com/foundriesio/fioconfig/pull/45/commits/348421ef08fb4d9e062b06813387d0b565d19e32

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:100]: I'm still on the fence with this Write function. Part of me wants to only allow writing to `<prefix>/sota.toml` and error out if a more significant file would override it. But you could also try and keep the write and make it a little smarter to check for read-only errors and finally have some sort of "/etc/z99-fioconfig-hack.toml" as a fail-safe.

I'll think more on it, but I'm open to some alternatives if you have any

File: internal/app_config.go
Comment (vkhoroz) [inline internal/app_config.go:41]: Nice, thanks.

**Review by vkhoroz — APPROVED** [review]
Ship It!

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:41]: we are bug for bug compatible with aktualizr :)

File: internal/app_config.go
Comment (mike-sul) [reply internal/app_config.go:100]: > Foremost, @doanac I'm fine with your justification.
> 
> Unrelated to this specific change, but related in general.
> 
> I guess some of our users would want to put their `sota.toml` of a read-only partition. For them, we still need to provide an ability to rotate the client certificate.
> 
> One such way is to store the sota config client cert paths in a separate file. So, maybe, we need to consider making that an option in e.g. `lmp-device-register`?

I am not sure it will be easy to implement it. aklite/libaktualizr makes too many assumptions about the certs and sota.toml location. We also have this https://github.com/foundriesio/fioconfig/blob/1b1929bff291fa440e7efe7a29dfed9625b60ade/contrib/aktualizr-toml-update#L24 which assumes that sota.toml is writable.

Comment (mike-sul) [reply internal/app_config.go:100]: If `import.tls_*` are set in a read-only file it kind of assumes that a customer wouldn't like to support cert rotation. Therefore, we need to make it impose the requirement the certs' attributes should be defined in a writable `*.toml` file.

Now, should  the `import.tls_*` and `tls.*_source` attributes be defined only in `sota.toml` or they can be defined in one or more `*.toml` in `/etc`?

IMO, having those attributes defined in more than one `*.toml` file can create lots of confusion and lead to issues.
So, IMO, I would restrict writing the cert related attributes only to one file, either `sota.toml` or maybe a new one in `/etc`, not sure the later option is practical though.

In general, these generic `*.toml` mechanism is too generic :). There are certain critical parameters that should be defined only in a one `*.toml` file and we shouldn't allow overriding them. For example, does it make sense to allow overriding URLs specified in `sota.toml`?

I would say that this generic mechanism is a great thing for overriding/customizing configuration at the build time (yocto recipes in meta-subs), let's call it a static configuration.

When it comes to dynamic (re-)configuration, overriding in generic way is too dangerous. It has to be restricted and specific, as we do with `z-50-fioctl.toml`.

Comment (doanac) [reply internal/app_config.go:100]: What complicates this is that we have a generic platform that people can do almost anything with. Just the "single sota.toml" constraint we'd like: A customer might introduce a `/usr/lib/sota/conf.d/sota.toml` file they intend for global common stuff *and* have `/var/sota/sota.toml`. Or mind blown: `/var/sota-customized-for-no-good-reason/sota.toml`. So just knowing which sota.toml we should write to isn't foolproof.

Comment (doanac) [reply internal/app_config.go:100]: And the parsing order doesn't give us clues. i.e. the first entry by default is /usr/lib and not /var/sota.

Comment (mike-sul) [reply internal/app_config.go:100]: I think we need some kind of top level config which is defined solely in yocto (factory's lmp-manifest, meta-subs). This top-level config will define the things like this, e.g.

1. A list of dirs/files containing `*.toml` snippets for aklite config.
2. A path to the `*.toml` file that contains mTLS/certs related config.

Comment (doanac) [reply internal/app_config.go:100]: k. i'm revisiting this finally. I think I have a compromise. The interface to update TOML settings could have 2 new things added:

1. If the file we find to write values to isn't writeable, then we move on and try to find a more significant file that is writable.
2. Create some type of "Can Update Config" method that takes a set of keys. This will allow us to test up front before trying to do the entire key rotation.

Comment (mike-sul) [reply internal/app_config.go:100]: I think it should work.

As for the item 1. If more significant writable file is not found then the method/function will return an error?

Comment (doanac) [reply internal/app_config.go:100]: yep. that's my thinking

**Comment by doanac** [comment]
@mike-sul @vkhoroz - take a look at this new approach and see what you think

File: internal/app_config.go
Comment (vkhoroz) [inline internal/app_config.go:153]: IIUC we only use this for cert rotation ATM, so at least one key is always present in at least one config file.

But, as this is a generic method, if no file contains a key - it does not mean that we cannot write that key to some file.
In fact, we may write to any file which is not `z-50-fioctl.toml` in this case.

File: go.mod
Comment (vkhoroz) [inline go.mod:3]: May we do this as a separate PR?

This one is already overloaded.

File: internal/app.go
Comment (vkhoroz) [inline internal/app.go:234]: IIUC this is an unrelated cleanup, right?
May it go on its own too?

**Review by vkhoroz — APPROVED** [review]
This new approach is more complex, but it makes things more robust.
So, I think it is a challenge we need to undertake.

I also proposed one more use case to improve it, which is not critical.

I'd really like @mike-sul to cross-check this.

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:153]: I think we need to be more explicit - also how do you determine the new file name? You have to guess or create one that sorts lexicographically after ever other file we might be parsing.

File: internal/app.go
Comment (doanac) [reply internal/app.go:234]: 1.22 complained so I bundled it with 1.22 changes

File: internal/app_config.go
Comment (vkhoroz) [reply internal/app_config.go:153]: I did not mean to create a new file.

What I meant is that given files as below, and trying to write `foo=bar`, we can write to either of those files.
```
sota1.toml:
  some=value

sota2.toml:
  other=value
```

File: internal/app.go
Comment (vkhoroz) [reply internal/app.go:234]: This is what I suspected, and this is what distracts from the _real thing_.
That's why I'm asking to extract the 1.22 upgrade (plus this cleanup) to a separate PR.

Comment (doanac) [reply internal/app.go:234]: https://github.com/foundriesio/fioconfig/pull/47

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:153]: I think I'd rather change the name of the method to something like `UpdateKeys` or just making it a private function explicitly for key rotation. fioconfig shouldn't be *adding* configuration

Comment (vkhoroz) [reply internal/app_config.go:153]: > change the name of the method ... making it a private function explicitly for key rotation

Yes, let's do it.
It is better to be explicit about what this function is about.

If we _later_ have a necessity to reuse this code, then we will _later_ decide how to do that.

Comment (vkhoroz) [reply internal/app_config.go:153]: That said, if that would take more than 5 minutes of your time, I'm fine with a one-line comment here like e.g. `If the key is absent in all config files - we don't try to add it.`
I think that would be explicit enough.

Comment (doanac) [reply internal/app_config.go:153]: rebased on latest main and force pushed this change to filter into each commit of the series

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:50]: I wonder if we need to check if `st.Name()` has the `.toml` extension?

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:83]: Not sure, if it is really critical at the moment. Sometimes, e.g. the `compose_apps` case, it is important to distishuishe between the case when the specified key is missing and when it is present in the sota conf but its value is `""`.

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go:50]: In this case the file was explicitly set by the caller. e.g. something like `fioconfig -c /var/sota.toml-andy checkin`. I think we should honor that.

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go]: Do we really need to make it absolutely atomically provided that aklite will re-read the conf onaly on restart? Probably for the case when a device is turned of in the middle of the writing the `keyVals` across many files.

Comment (doanac) [reply internal/app_config.go]: > Probably for the case when a device is turned of in the middle of the writing the keyVals across many files.

That's why we have to do it atomically. If we were doing a key rotation and only updated one of two files - the device would be a bad state.

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:89]: `in sota.toml` is a bit confusing IMO, since the "get" code tries to find and get the key value across all configuration files.

Comment (doanac) [reply internal/app_config.go:89]: ah - good catch. that's copy/paste from the old stuff.

File: internal/app_config.go
Comment (mike-sul) [inline internal/app_config.go:142]: It looks like we have to decrease `j` starting from `i-1` here?

**Comment by mike-sul** [comment]
It is a bit odd that a certain `key=value` can be moved from the current *.toml file to some more significant one as result of the batched config update. On the other side it looks like a good trade off for the atomicity.

**Comment by doanac** [comment]
> It is a bit odd that a certain key=value can be moved from the current *.toml file to some more significant one as result of the batched config update.

It's highly unlikely to ever happen. It's also highly unlikely a device would crash while writing to two different files. This is the edge case of an edge case.

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go]: https://github.com/foundriesio/fioconfig/pull/45/commits/71a1cea2b191a3f410892f8c758935699c4f2c69

File: internal/app_config.go
Comment (doanac) [reply internal/app_config.go]: https://github.com/foundriesio/fioconfig/pull/45/commits/1c144006b76688346a22d73de284202739de9bf8

**Review by mike-sul — APPROVED** [review]

**Comment by doanac** [comment]
force push with squashed commits. will merge when ci passes

**Pull request merged**
